### DISCUSSION TITLE
back realtion filtering

### DISCUSSION THREAD

**Discussion by chandra447** [pocketbase/pocketbase#5112] [Q&A]
Hey,
I am trying to do the back filtering i have 2 collections employees and attendance.
![image](https://github.com/pocketbase/pocketbase/assets/39896565/9fd6c776-d419-4867-907d-b61f8aa5e472)

   This is my current implementation but it doesn't get and expand records, but when i remove the date filter i get all the records.
   
   
I have also tried adding the filtering to the filter section as below but it does only the inner join but doesnot get all the records which doesn't have back relation record.
![image](https://github.com/pocketbase/pocketbase/assets/39896565/cd622b09-c79e-44e8-9ac3-9576da77b60c)

Any suggestion or reference would help 🙏 
Thanks

**Comment by ganigeorgiev** [comment]
It is not clear what you are trying to do.

The `expand` syntax shown in the first screenshot is invalid. `expand` doesn't support filtering and sorting.

The `filter` apply **only** on the main/queried collection (in your case `employees`), aka. by using `attendance_via_employee.date > ...` you are requesting only `employees` that has attendance relations matching your criteria.

If you want to filter the `attendances` records you'll have to send a separate individual request to the `attendances`  collection.

If you need further help, please elaborate a little more on the desired behavior and your data structure.

**Reply by chandra447** [reply]
@ganigeorgiev 
what i am trying to do is, i have added the SQL query which i want to achieve
`select e.*,a.isPresent as expand from 
employees as e left join attendance as a on a.employee=e.employee
where a.date>='2024-06-21 00:00:00'`

**Reply by ganigeorgiev** [reply]
Sorry I don't think I will be able to  help you as the above doesn't make sense to me. Please don't treat the discussion as llm chat and try to elaborate more on your collections structure and the desired output (see also [XY problem](https://xyproblem.info/)). If you are not able to clearly express what you are trying to do then I don't think anyone will be able to assist you.

Again, `filter` filters only the main/queried collection (from your initial post - `employees`). There is no support to apply filter on the expanded records. You'll have to send separate request(s).

Additionally, note that the full datetime format is `Y-m-d H:i:s.uZ`, or in other words the date filter should be against `2024-06-21 00:00:00.000Z`.

**Reply by chandra447** [reply]
hi @ganigeorgiev ,
i apoligize for my poor explanation of what i am trying to solve.
I have this ER diagram to better explain the current schema
![image](https://github.com/pocketbase/pocketbase/assets/39896565/77395bab-fad7-4d09-aa38-ce9c3391450a)
what i am trying to do is get all the employees and expand on the attendance collection where `Attendance.date>=currentDate(replacingT when used new Date)`

Based on the explanation above, i am able to understand that it is not possible to filter on expand.
Correct me f my understanding is wrong.
Thank you 🙏

**Reply by ganigeorgiev** [reply]
This is really frustrating. Please understand that I have no context about your collections and by using different names in different examples makes it even more confusing to provide a working solution. Is the name of the relation field in the Attendence collection "employeeId" or just `employee`? Is the name of the employees and attendacees collection singlular or plural?
The `expand` result will also depend on the View API rule of the related collection and by having no information about this I don't know if your problem is related to this or something else.

The `filter` doesn't apply on the expanded relations. I don't know how to explain it more clear than this. Again:

> The `filter` apply only on the main/queried collection (in your case employees), aka. **by using attendance_via_employee.date > ... you are requesting only employees that has attendance relations matching your criteria**.

I'll try with an example one last time and will stop engage to avoid further confusion.

Let's assume that the name of the collections are `employees` and `attendances` and that the relation field is named "employee". We have the following data:

- EmployeeA
   - attendance1 (2024-06-15 10:20:00.123Z)
   - attendance2 (2024-06-17 10:40:00.123Z)
   - attendance3 (2024-06-23 10:40:00.123Z)

- EmployeeB
   - attendance4 (2024-06-21 10:20:00.123Z)
   - attendance5 (2024-06-22 10:40:00.123Z)
   
- EmployeeC
   - attendance6 (2024-06-21 10:20:00.123Z)
   - attendance7 (2024-06-24 10:40:00.123Z)


By using:

```js
const employees = await pb.collection("employees").getFullList({
    filter: "attendances_via_employee.date >= 2024-06-21 00:00:00.000Z",
    expand: "attendances_via_employee"
})
```

The above will match:

- EmployeeB
   - attendance4 (2024-06-21 10:20:00.123Z)
   - attendance5 (2024-06-22 10:40:00.123Z)
   
- EmployeeC
   - attendance6 (2024-06-21 10:20:00.123Z)
   - attendance7 (2024-06-24 10:40:00.123Z)

---

Now if you prefix the operator with `?` (aka. any/at-least-one):

```js
const employees = await pb.collection("employees").getFullList({
    filter: "attendances_via_employee.date ?>= 2024-06-21 00:00:00.000Z",
    expand: "attendances_via_employee"
})
```

This will return all 3 employees (EmployeeA, EmployeeB, EmployeeC) each with all of their attendancees expanded (including the attendance1 and attendance2 because attendance3 matches the filter).

- EmployeeA
   - attendance1 (2024-06-15 10:20:00.123Z)
   - attendance2 (2024-06-17 10:40:00.123Z)
   - attendance3 (2024-06-23 10:40:00.123Z)

- EmployeeB
   - attendance4 (2024-06-21 10:20:00.123Z)
   - attendance5 (2024-06-22 10:40:00.123Z)
   
- EmployeeC
   - attendance6 (2024-06-21 10:20:00.123Z)
   - attendance7 (2024-06-24 10:40:00.123Z)

**Reply by chandra447** [reply]
First of all thank you for holding up with my confused explanation. I messed up the names in the ER Diagram, but let us continue with your example provided,Is there any possibility to get **employeeA** as well considering the first suggestion you have provided. 
Though EmployeeA doesn't have any records, how can i get EmployeeA as well when i query but with no `expand key`
Should i do a seperate request and append? or is there a possibility?
```javascript
const employees = await pb.collection("employees").getFullList({
    filter: "attendances_via_employee.date >= 2024-06-21 00:00:00.000Z",
    expand: "attendances_via_employee"
})
```
